sdk-cpp/src/internal/com/do_download_property_internal.cpp

```diff
@@ -19,77 +19,85 @@ std::wstring UTF8toWstr(const char* str, size_t cch = 0)
 
     if (cch == 0)
     {
-        return std::wstring();
+        wstr = std::wstring();
     }
 
     std::vector<wchar_t> dest(cch * 4);
-    const UINT result = MultiByteToWideChar(CP_UTF8, 0, str, static_cast<int>(cch), dest.data(), static_cast<int>(dest.size()));
+    const uint32_t result = MultiByteToWideChar(CP_UTF8, 0, str, static_cast<int>(cch), dest.data(), static_cast<int>(dest.size()));
     if (result == 0)
     {
-        throw std::exception();
+        return E_FAIL;
     }
-    return std::wstring(dest.data(), result);
+    wstr = std::wstring(dest.data(), result);
+    return S_OK;
 }
 
-CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(const std::string& val)
+int32_t CDownloadPropertyValueInternal::Init(const std::string& val) noexcept
 {
     V_VT(&_var) = VT_BSTR;
 
-    std::wstring wval = UTF8toWstr(val.c_str());
+    std::wstring wval;
+    auto hr = UTF8toWstr(wval, val.c_str());
+    RETURN_IF_FAILED(hr);
 
     BSTR bstr = SysAllocString(wval.c_str());
     if (bstr == nullptr)
     {
-        throw std::bad_alloc();
+        hr = DO_ERROR_FROM_STD_ERROR(std::errc::not_enough_memory);
     }
     V_BSTR(&_var) = bstr;
+
+    return S_OK;
 };
 
-CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(uint32_t val)
+int32_t CDownloadPropertyValueInternal::Init(uint32_t val) noexcept
 {
     V_VT(&_var) = VT_UI4;
     V_UI4(&_var) = val;
+    return S_OK;
 };
 
-CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(uint64_t val)
+int32_t CDownloadPropertyValueInternal::Init(uint64_t val) noexcept
 {
     V_VT(&_var) = VT_UI8;
     V_UI8(&_var) = val;
+    return S_OK;
 };
 
-CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(bool val)
+int32_t CDownloadPropertyValueInternal::Init(bool val) noexcept
 {
     V_VT(&_var) = VT_BOOL;
     V_BOOL(&_var) = val ? VARIANT_TRUE : VARIANT_FALSE;
+    return S_OK;
 };
 
-CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(std::vector<unsigned char>& val)
+int32_t CDownloadPropertyValueInternal::Init(std::vector<unsigned char>& val) noexcept
 {
-    throw errc::e_not_impl;
+    return static_cast<int32_t>(errc::e_not_impl);
 };
 
-CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(const download_property_value::status_callback_t& val)
+int32_t CDownloadPropertyValueInternal::Init(const download_property_value::status_callback_t& val) noexcept
 {
+    V_VT(&_var) = VT_EMPTY;
```

Comment:
>V_VT(&_var) = VT_EMPTY; [](http://example.com/codeflow?start=4&length=23)

Why not use VariantInit? Standard/conventional way.

Reply:
Since the std::vector<unsigned char> one also requires this, I suggest implementing the default constructor and calling VariantInit there. The constructor on the rest side can use = default in its .cpp file.